### PR TITLE
fix: #1065 Add custom values to slider

### PR DESCRIPTION
**Pull request by metaboulie**
Screenshot 2024-04-25 at 19 02 47

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -96,6 +96,7 @@
     "lodash-es": "^4.17.21",
     "lucide-react": "^0.338.0",
     "lz-string": "^1.5.0",
+    "mathjs": "^12.4.2",
```
Comment (mscolnick) [inline frontend/package.json:99]: i think youll want to run `pnpm install` in `cd frontend`, and commit the changes to the lockfile

File: frontend/src/utils/numbers.ts [hunk 2]
```diff
@@ -1,7 +1,31 @@
...
+  if (Number.isNaN(value)) {
+    return "NaN";
+  }
+  if (!Number.isFinite(value)) {
+    return value > 0 ? "Infinity" : "-Infinity";
+  }
+
+  return Math.trunc(value) === 0
+    ? // No integer part, use scientific notation
+      format(value, { notation: "auto", precision: 2 })
```
Comment (mscolnick) [inline frontend/src/utils/numbers.ts:25]: the browser has this built in now. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat#scientific

would that work instead and avoid the extra dep? we can lowercase the `e` aswell to match this formatting. 


```ts
new Intl.NumberFormat("en-US", {
  notation: "scientific",
}).format(987654321);
```

File: tests/_plugins/ui/_impl/test_input.py [hunk 3]
```diff
@@ -75,6 +84,18 @@ def test_slider_invalid_bounds() -> None:
     assert "Invalid bounds" in str(e.value)
 
 
+def test_slider_exclusive_args() -> None:
```
Comment (mscolnick) [inline tests/_plugins/ui/_impl/test_input.py:87]: can we add a test for empty `steps` (not sure how we want to handle) and invalid bounds (value is not in steps)

File: tests/_plugins/ui/_impl/test_input.py [hunk 4]
```diff
@@ -114,6 +137,13 @@ def test_range_slider_init() -> None:
         # cast to floats
         assert isinstance(num, float)
 
+    slider = ui.range_slider(steps=[1, 3, 6, 10, 17, 20], value=[3, 17])
```
Comment (mscolnick) [inline tests/_plugins/ui/_impl/test_input.py:140]: can we add a test for empty `steps` (not sure how we want to handle) and invalid bounds (value is not in steps)

File: frontend/package.json (hunk 1 above)
Comment (mscolnick) [reply frontend/package.json:99]: maybe can remove this with the suggestion below

File: frontend/src/utils/numbers.ts (hunk 2 above)
Comment (metaboulie) [reply frontend/src/utils/numbers.ts:25]: I think there is a bug for this, the slider would disappear after some dragging, so I retain to `mathjs`

File: frontend/package.json (hunk 1 above)
Comment (metaboulie) [reply frontend/package.json:99]: I've updated the lockfile, but don't know whether it's right

**Review by mscolnick — APPROVED** [review]
awesome! this is really great
